File: modules/lindstrom-hsm-communications/android/src/main/java/expo/modules/lindstromhsmcommunications/LindstromHsmCommunicationsModule.kt
package expo.modules.lindstromhsmcommunications

import com.google.gson.Gson
import com.hagleitner.hsmdevicecommunications.*
import com.hagleitner.hsmdevicecommunications.bleconnectionprotocol.builder.BleConnectionProtocolMessageBuilder
import com.hagleitner.hsmdevicecommunications.bleconnectionprotocol.builder.ReadTLV
import com.hagleitner.hsmdevicecommunications.bleconnectionprotocol.builder.WriteTLV2
import com.hagleitner.hsmdevicecommunications.bleconnectionprotocol.parser.BleConnectionProtocolMessage
import com.hagleitner.hsmdevicecommunications.bleconnectionprotocol.parser.ResourceIdentifier
import com.hagleitner.hsmdevicecommunications.bleconnectionprotocol.parser.enums.DataType
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.Base64

class LindstromHsmCommunicationsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("LindstromHsmCommunications")

    // Parse BLE advertisement without performing decryption
    Function("parseBleAdvertisementWithoutDecryption") { manufacturerData: String ->
      try {
        val byteArray = Base64.getDecoder().decode(manufacturerData)
        val result = parseBleAdvertisementWithoutDecryption(byteArray)
        return@Function Gson().toJson(result)
      } catch (e: IllegalArgumentException) {
        println("Invalid Base64 string: ${e.message}")
        return@Function null
      }
    }

    // Decrypt resource values
    Function("decryptResourcesValues") { encryptedResourceValuesHex: String, decryptionKey: String ->
      try {
        val result: ByteArray = decryptResourcesValues(encryptedResourceValuesHex, decryptionKey)
        return@Function Gson().toJson(result)
      } catch (e: IllegalArgumentException) {
        println("Error while decoding: ${e.message}")
        return@Function null
      }
    }

    // Get resource values as integer
    Function("getResourcesValuesAsInt") { decryptedByteArray: String, offset: Int, length: Int ->
      try {
        val byteArray = Base64.getDecoder().decode(decryptedByteArray)
        return@Function getResourcesValuesAsInt(byteArray, offset, length)
      } catch (e: IllegalArgumentException) {
        println("Error in getResourcesValuesAsInt: ${e.message}")
        return@Function null
      }
    }

    // Build encrypted single read message
    Function("buildEncryptedReadMessage") { resourceType: Int, instance: Int, appKey: String ->
      try {
        val resource = ResourceIdentifier(resourceType.toUInt().toUShort(), instance.toUInt().toUShort())
        val (encryptedReadCommand, tlvId) = BleConnectionProtocolMessageBuilder.buildEncryptedReadMessage(
          ReadTLV(resource = resource),
          key = appKey
        )
        return@Function mapOf(
          "encryptedMessage" to Base64.getEncoder().encodeToString(encryptedReadCommand),
          "tlvId" to tlvId.toInt(),
          "resourceType" to resourceType,
          "instance" to instance
        )
      } catch (e: Exception) {
        println("Error in buildEncryptedReadMessage: ${e.message}")
        return@Function null
      }
    }

    // Build encrypted multiple read messages
    Function("buildEncryptedReadMessages") { resources: List<Map<String, Int>>, appKey: String ->
      try {
        val tlvs = resources.map { resourceMap ->
          val resourceType = (resourceMap["resourceType"] as? Number)?.toInt()
            ?: throw IllegalArgumentException("resourceType is missing or not a valid number")
          val instance = (resourceMap["instance"] as? Number)?.toInt()
            ?: throw IllegalArgumentException("instance is missing or not a valid number")


          val resource = ResourceIdentifier(resourceType.toUInt().toUShort(), instance.toUInt().toUShort())
          ReadTLV(resource = resource)
        }

        val (encryptedMessage, tlvIds) = BleConnectionProtocolMessageBuilder.buildEncryptedMessage(
          tlvs = tlvs,
          key = appKey
        )

        return@Function mapOf(
            "encryptedMessage" to Base64.getEncoder().encodeToString(encryptedMessage),
            "tlvIds" to tlvIds.map { it.toInt() } // easier for JS to work with
        )
      } catch (e: Exception) {
        println("Error in buildEncryptedReadMessages: ${e.message}")
        return@Function null
      }
    }

    // Build encrypted write message
    Function("buildEncryptedWriteMessage") {
      resourceType: Int, instance: Int, appKey: String, value: String, dataType: Int ->

      try {
        val type = DataType.entries.find { it.ordinal == dataType }
          ?: throw IllegalArgumentException("Data type not found")

        val resource = ResourceIdentifier(resourceType.toUInt().toUShort(), instance.toUInt().toUShort())
        val (encryptedWriteCommand, tlvId) = BleConnectionProtocolMessageBuilder
          .buildEncryptedMessage(
            listOf(
              WriteTLV2(
                resource = resource,
                value = value,
                dataType = type
              )
            ),
            key = appKey
          ).let { (encrypted, ids) -> encrypted to ids.first() }

        return@Function mapOf(
          "encryptedMessage" to Base64.getEncoder().encodeToString(encryptedWriteCommand),
          "tlvId" to tlvId.toInt(),
          "resourceType" to resourceType,
          "instance" to instance
        )
      } catch (e: Exception) {
        println("Error in buildEncryptedWriteMessage: ${e.message}")
        return@Function null
      }
    }

    // Interpret received value
    Function("interpretReceivedValue") { value: String, appKey: String ->
      try {
        val byteArray = Base64.getDecoder().decode(value)
        val protocolMessage = BleConnectionProtocolMessage.fromEncryptedByteArray(byteArray, appKey)

        val receivedData = protocolMessage.resourceMessages.map { resourceMessage ->
          mapOf(
            "value" to resourceMessage.resourceDefinition.value.boxedValue.toString(),
            "tlvId" to resourceMessage.id.id.toInt()
          )
        }

        return@Function receivedData
      } catch (e: IllegalArgumentException) {
        println("Error in interpretReceivedValue: ${e.message}")
        return@Function null
      }
    }

    // Get data types
    Function("getDataTypes") {
      return@Function DataType.values().associate { it.ordinal to it.name }
    }
  }
}

File: modules/lindstrom-hsm-communications/src/LindstromHsmCommunicationsModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import {
  BuildEncryptedReadMessage,
  BuildEncryptedReadMessages,
  BuildEncryptedWriteMessage,
  ByteArray,
  DecryptedMessagePayload,
  EncryptedMessagePayload,
} from './LindstromHsmCommunications.types';

/** Native module interface for HSM Device Communications. */
declare class LindstromHsmCommunicationsModule extends NativeModule {
  /**
   * Parses a BLE advertisement without performing decryption.
   * @param manufacturerData - The manufacturer data in hexadecimal format.
   * @returns A JSON string representation of the parsed BLE advertisement (BleAdvertisementV2 format, parsed on the JavaScript side).
   */
  parseBleAdvertisementWithoutDecryption(manufacturerData: string): string;

  /**
   * Decrypts resource values using the provided decryption key.
   * @param encryptedResourceValuesHex - The encrypted resource values in hex format.
   * @param decryptionKey - The decryption key.
   * @returns The decrypted resource values as a byte array.
   */
  decryptResourcesValues(
    encryptedResourceValuesHex: string,
    decryptionKey: string
  ): ByteArray;

  /**
   * Extracts resource values as an integer from a decrypted byte array.
   * @param decryptedByteArray - The decrypted byte array as a string.
   * @param offset - The offset to start reading from.
   * @param length - The number of bytes to read.
   * @returns The extracted integer value.
   */
  getResourcesValuesAsInt(
    decryptedByteArray: string,
    offset: number,
    length: number
  ): number;

  /**
   * Builds an encrypted read message.
   * @param resourceType - The resource type.
   * @param instance - The instance identifier.
   * @param appKey - The application key.
   * @returns The encrypted read message payload.
   */
  buildEncryptedReadMessage(
    resourceType: BuildEncryptedReadMessage['resourceType'],
    instance: BuildEncryptedReadMessage['instance'],
    appKey: BuildEncryptedReadMessage['appKey']
  ): {
    encryptedMessage: EncryptedMessagePayload['encryptedMessage'];
    tlvId: EncryptedMessagePayload['tlvId'];
    resourceType: EncryptedMessagePayload['resourceType'];
    instance: EncryptedMessagePayload['instance'];
  };

  /**
   * Builds an encrypted read message for a list of resources.
   * @param resources - An array of resource types and instances.
   * @param appKey - The application key.
   * @returns The encrypted read message payload and the list of the TLV IDs for each resource.
   */
  buildEncryptedReadMessages(
    resources: BuildEncryptedReadMessages['resources'],
    appKey: BuildEncryptedReadMessages['appKey']
  ): {
    encryptedMessage: EncryptedMessagePayload['encryptedMessage'];
    tlvIds: EncryptedMessagePayload['tlvId'][];
  };

  /**
   * Builds an encrypted write message.
   * @param resourceType - The resource type.
   * @param instance - The instance identifier.
   * @param appKey - The decryption key.
   * @param value - The value to write.
   * @param dataType - The data type of the value.
   * @returns The encrypted write message payload.
   */
  buildEncryptedWriteMessage(
    resourceType: BuildEncryptedWriteMessage['resourceType'],
    instance: BuildEncryptedWriteMessage['instance'],
    appKey: BuildEncryptedWriteMessage['appKey'],
    value: BuildEncryptedWriteMessage['value'],
    dataType: BuildEncryptedWriteMessage['dataType']
  ): {
    encryptedMessage: EncryptedMessagePayload['encryptedMessage'];
    tlvId: EncryptedMessagePayload['tlvId'];
    resourceType: EncryptedMessagePayload['resourceType'];
    instance: EncryptedMessagePayload['instance'];
  };

  /**
   * Interprets a received value using the provided decryption key.
   * @param value - The received value as a string.
   * @param appKey - The decryption key.
   * @returns The interpreted value and its TLV ID.
   */
  interpretReceivedValue(
    value: string,
    appKey: string
  ): {
    value: DecryptedMessagePayload['value'];
    tlvId: DecryptedMessagePayload['tlvId'];
  }[];

  /**
   * Retrieves the supported data types.
   * @returns An array of supported data types as strings.
   */
  getDataTypes(): string[];
}

// Load the native module object from the JSI.
export default requireNativeModule<LindstromHsmCommunicationsModule>(
  'LindstromHsmCommunications'
);
